docs/cryptodoc/src/05_08_ml_dsa.rst

```diff
@@ -0,0 +1,443 @@
+It is not advisable to use any other variant than the ones specified in
+[FIPS-204]_.
+
+.. _pubkey/ml_dsa/keygen:
+
+Key Generation
+--------------
+
+Generating a fresh ML-DSA key pair as specified in [FIPS-204]_ Section 5.1
+Algorithm 1, is available in the constructor of :srcref:`Dilithium_PrivateKey
+<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:403|Dilithium_PrivateKey>`.
+This mostly delegates the actual key generation to the internal function
+:srcref:`expand_keypair
+<src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:665|expand_keypair>`
+that follows [FIPS-204]_ Section 6.1 Algorithm 6.
+
+.. admonition:: Dilithium_PrivateKey::Dilithium_PrivateKey / Dilithium_Algos::expand_keypair
+
+   **Input:**
+
+   - ``rng``: random number generator
+   - ``mode``: ML-DSA parameter set descriptor
+
+   **Output:**
+
+   - ``sk``: private signing key
+   - ``pk``: public verification key
+
+   **Steps:**
+
+   1. Generate a random 32-byte seed :math:`\xi` using ``rng``
+   2. :math:`(\rho, \rho', K) = H(\xi)` (32, 64, and 32 bytes respectively)
+   3. Sample matrix :math:`\hat{A}` from :math:`\rho` using ``expand_A``
+   4. Sample vectors :math:`s_1` and :math:`s_2` from :math:`\rho'` using ``expand_s``
+   5. Calculate :math:`(t_1, t_0)` from :math:`\hat{A}`, :math:`s_1`, and :math:`s_2` using ``compute_t1_and_t0`` (see :srcref:`here <src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:310|compute_t1_and_t0>`)
+
+      1. :math:`t = NTT^{-1}(\hat{A} \cdot NTT(s_1)) + s_2`
+      2. :math:`(t_1, t_0) = power2round(t)` (see :srcref:`here <src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:746|power2round>`)
+
+   6. :math:`pk = (\rho, t_1)` and :math:`sk = (\xi, K, s_1, s_2, t_0)`
+
+   **Notes:**
+
+   - Step 1 corresponds to [FIPS-204]_ Algorithm 1
+   - Steps 2-5 correspond to [FIPS-204]_ Algorithm 6
+   - Step 6 returns the key pair in Botan's internal representation. The
+     encoding and hashing of the encoded public key are done later and on
+     demand.
+
+
+.. _pubkey/ml_dsa/signing:
+
+Signing
+-------
+
+Signature generation as specified in [FIPS-204]_ Algorithms 2 and 7 are
+implemented in :srcref:`Dilithium_Signature_Operation::sign
+<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:153|sign>` with the
+preparation of the message representative :math:`\mu` being done in
+:srcref:`DilithiumMessageHash
+<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.
+
+.. admonition:: Dilithium_Signature_Operation::sign
+
+   **Input:**
+
+   - ``sk``: private signing key, with :math:`\hat{A}` and :math:`\hat{s_1}`, :math:`\hat{s_2}`, and :math:`\hat{t_0}` in NTT domain
+   - ``M``: message to be signed
+
+   **Output:**
+
+   - ``signature``: the valid signature
+
+   **Steps:**
+
+   1. Calculate the message representative :math:`\mu = H(sk.tr \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)
```

Comment:
I guess instead of `sk.tr`this should be `H(pk)`or similar since the hashing is done on demand as stated in the keygen? Or is it a member of `sk` that is computed on first use?

Reply:
The public key is not an input variable for signing. And the private key is defined to contain `tr = H(pk)`. But you're right, this is not obvious, I'll note that `sk.tr = H(pk)` in the `Input` section.